sycl/test/extensions/properties/new_properties.cpp

```diff
@@ -0,0 +1,242 @@
+  static_assert(!pl2.has_property<property1>());
+
+  static_assert(pl1.has_property<property2>());
+  static_assert(!pl2.has_property<property2>());
+
+  static_assert(pl1.has_property<property3>());
+  static_assert(pl2.has_property<property3>());
+}
+} // namespace test
+
+namespace bench {
+template <int N> struct property : named_property_base<property<N>> {
+  static constexpr int value() { return N; }
+};
+
+template <int... N> void test(std::integer_sequence<int, N...>) {
+  properties pl{property<N>{}...};
+  static_assert((pl.template has_property<property<N>>() && ...));
+  static_assert(
+      ((pl.template get_property<property<N>>().value() == N) && ...));
+}
+} // namespace bench
+
+namespace test_operator_plus {
+template <int N> struct property : named_property_base<property<N>> {};
+
+constexpr properties pl1{property<1>{}, property<2>{}, property<3>{}};
+constexpr properties pl2 = pl1 + properties{property<4>{}};
+static_assert(!pl1.has_property<property<4>>());
+static_assert(pl2.has_property<property<2>>());
+static_assert(pl2.has_property<property<4>>());
+} // namespace test_operator_plus
+
+namespace test_compile_prop_in_runtime_list {
+template <int N>
+struct ct_prop
+    : named_property_base<ct_prop<N>,
+                          detail::property_key_value_template<ct_prop>> {
+  static constexpr auto value() { return N; }
+};
+struct rt_prop : named_property_base<rt_prop> {
+  rt_prop(int N) : x(N) {}
+
+  int x;
+
+  constexpr auto value() { return x; }
+};
+void test() {
+  int x = 42;
+  properties pl{ct_prop<42>{}, rt_prop{x}};
+  constexpr auto p = pl.get_property<ct_prop>();
+  static_assert(std::is_same_v<decltype(p), const ct_prop<42>>);
+  static_assert(p.value() == 42);
+}
+} // namespace test_compile_prop_in_runtime_list
+
+namespace test_static_get_property {
+  struct ct_prop : named_property_base<ct_prop> {};
+  struct rt_prop : named_property_base<rt_prop> {
+    int x;
+  };
+  void test() {
+    properties pl{ct_prop{}, rt_prop{}};
+    constexpr auto c = decltype(pl)::get_property<ct_prop>();
+    auto r = pl.get_property<rt_prop>();
+  }
+}
+
+namespace properties_validation_example {
+struct prop : named_property_base<prop> {};
+struct prop2 : named_property_base<prop2> {};
+
+template <typename> struct foo;
+template <typename... property_tys>
+struct foo<properties<detail::properties_type_list<property_tys...>>> {
+  template <typename... key_tys> static constexpr bool bar(key_tys...) {
+    return ((sycl::detail::check_type_in_v<typename property_tys::key_t,
+                                           key_tys...> &&
+             ...));
+  }
+};
+constexpr properties pl{prop{}};
+using ty = std::remove_const_t<decltype(pl)>;
+static_assert(foo<ty>::bar(detail::key<prop>()));
+static_assert(foo<ty>::bar(detail::key<prop>(), detail::key<prop2>()));
+static_assert(foo<empty_properties_t>::bar(detail::key<prop>()));
+static_assert(!foo<ty>::bar(detail::key<prop2>()));
```

Comment:
Let's continue the discussion from https://github.com/intel/llvm/pull/15899#discussion_r1820672750 here. @gmlueck , @Pennycook

Reply:
@gmlueck 's argument is that we want to be able to extend the set of allowed properties for a given API from multiple extensions. As such, it's preferred to have a trait that each property can specialize, instead of a single type list (as in @Pennycook's and my suggestions).